### ISSUE TITLE
Support static/shared variants of Boost libraries

### ISSUE THREAD

**Issue by ruslo** [ruslo/hunter#130]
For now only static version can be built, `BUILD_SHARED_LIBS=ON` has no effect.
##

**Comment by ruslo** [comment]
For now removed: https://github.com/ruslo/hunter/commit/4618d4a3ac537d7cea02e326caa716f1434b4d57

**Comment by ruslo** [comment]
Related:
* https://github.com/ruslo/hunter/issues/21
* https://github.com/ruslo/hunter/issues/146

**Comment by ruslo** [comment]
Remove unused script: https://github.com/ruslo/hunter/commit/c822f08d8c5451d0a63f80b766076002b1b4f46a

**Comment by LegalizeAdulthood** [comment]
I bumped into this last night in my Hunter presentation to Utah C++ Programmers.   I was able to quickly figure out what was going on, but only because I am pretty familiar with how to build boost.

This needs to be called out explicitly on the wiki page for boost and tell people that they have to compile all their code with `BOOST_ALL_NO_LIB=1`.  Otherwise, you get weird errors on Windows.

I also noticed that the boost libraries don't have the toolchain in their filename as I expected...

**Comment by pretyman** [comment]
I think if we create a compile definition for the boost compiled libraries, this issue would be transparent to the outside world:
```CMake
target_compile_definitions(...
    INTERFACE
      BOOST_ALL_NO_LIB=1
)
```
let's have @ruslo 's thoughts on this

**Comment by pretyman** [comment]
Spoke too soon, it seems to be there for VisualStudio: https://github.com/ruslo/hunter/blob/master/cmake/templates/BoostConfig.cmake.in#L39

**Comment by ruslo** [comment]
> I also noticed that the boost libraries don't have the toolchain in their filename as I expected...

@LegalizeAdulthood Not sure what does it mean. Also why it matters?

> This needs to be called out explicitly on the wiki page for boost

It's already on the wiki page: https://github.com/ruslo/hunter/wiki/pkg.boost#compatibility-mode

```cmake
if(MSVC)
  add_definitions(-DBOOST_ALL_NO_LIB=1)
endif()
```

Are you using Visual Studio?

Also as @pretyman noticed this definition added to imported targets like `Boost::system` automatically.

**Comment by LegalizeAdulthood** [comment]
@ruslo Oh, now I see where the information is buried.  But it needs to be just stated in plain English at the top of the page: "This only supplies static libraries."

As for the toolset not being in the filename, I don't know if Boost puts that in the filenames when you build the static libs, but it is definitely in the fiename when you build dynamic libs.  Since I normally build dynamic libs, that's what I've come to expect to see in the filename.  I was just making an observation.  Using CMake's `${Boost_LIBRARIES}` handles the appropriate filename, so it works properly.

**Comment by LegalizeAdulthood** [comment]
@pretyman The help for `target_compile_definitions` says that this is a list of definitions to use when compiling the target.  Adding that to the target for boost would just affect how the libraries are compiled when building boost itself. The preprocessor symbol `BOOST_ALL_NO_LIB` is to be used by clients of boost libraries.  (I am not sure if Boost also uses it internally to determine whether things are static or dynamic.)  When I've built boost, the static/dynamic options are passed into `b2` and it drives everything from a high-level; I am unsure of the exact low-level mechanics.

At any rate, we eventually need both static and dynamic library support in the boost recipe.  I was just commenting that it should be stated more prominently on the wiki page for the package.

**Comment by schuhschuh** [comment]
What about the INTERFACE option of ```target_compile_definitions``` ?

**Comment by ruslo** [comment]
> But it needs to be just stated in plain English at the top of the page: "This only supplies static libraries."

I think wiki page structured quite correctly. First came usage then bugs and workarounds. I've added this bug to the last section. Also I don't understand why does it matters. Users should do `hunter_add_package`/`find_package(Boost ...)`/`target_link_libraries(.... Boost::system)` and everything will just work.

**Comment by ruslo** [comment]
> Adding that to the target for boost would just affect how the libraries are compiled when building boost itself

@LegalizeAdulthood Not always. `INTERFACE` is for the client.

**Comment by ruslo** [comment]
> What about the INTERFACE option of target_compile_definitions ?

@schuhschuh This is already implemented for the imported targets like `Boost::system`.

**Comment by schuhschuh** [comment]
@ruslo Indeed :+1:, I had not read the other comments before. I only replied to the incorrect statement I maybe too quickly read and replied to in my inbox :-)

**Comment by LegalizeAdulthood** [comment]
Well, it doesn't just work, that's my point.

For instance, when I hunter-ified [this `CMakeLists.txt`](https://github.com/exercism/xcpp/blob/master/bob/CMakeLists.txt), by following the instructions on the main Hunter page, I got a link error attempting to find the dynamic library.  Knowing Boost as I do, I looked in the Hunter build area and saw only static libraries, so I knew to add `BOOST_ALL_NO_LIB=1` to my source and recompile.

Anyone who isn't already familiar with the guts of boost and uses the typical Boost build instructions when building it themselves, uses `build-type=complete` and gets both static and dynamic libraries.  When switching to Hunter, things change and you have to understand what is going on.  Even though *I* know what `BOOST_ALL_NO_LIB=1` means, I still didn't notice it buried down at the bottom of the wiki page.  Someone who is less familiar with boost isn't going to know what this means and to make things easier, we should just simply state that the package only provides static libraries.

`BOOST_ALL_NO_LIB` doesn't mean "use static libraries" either, it simply disables the automatic linking to shared libraries through Visual Studio.  If I have shared libraries I can still use `BOOST_ALL_NO_LIB` and specify the correct shared-library library myself when linking.  This is clearly spelled out in the [documentation](http://www.boost.org/doc/libs/1_58_0/boost/config/user.hpp).

**Comment by ruslo** [comment]
> For instance, when I hunter-ified this CMakeLists.txt, by following the instructions on the main Hunter page, I got a link error attempting to find the dynamic library

This is because you didn't follow instructions on wiki page. If you want to use imported targets you should use `find_package(Boost CONFIG)`. When you're using old-style `find_package(Boost MODULE)` you should add `set(Boost_USE_STATIC_LIBS ON)` and `add_definitions(-DBOOST_ALL_NO_LIB=1)` for MSVC.

> uses build-type=complete and gets both static and dynamic libraries

This violates main CMake style work with static/shared libs. Usually if you want static you run with `BUILD_SHARED_LIBS=OFF`, if you want dynamic - `BUILD_SHARED_LIBS=ON`.

**Comment by LegalizeAdulthood** [comment]
I'm not sure how `CONFIG` affects the whole setup, but you're right I didn't have that.

**Comment by ruslo** [comment]
> I'm not sure how CONFIG affects the whole setup

@LegalizeAdulthood With `MODULE` we will be looking for `FindBoost.cmake` (this is default). With `CONFIG` we will be looking for `BoostConfig.cmake`. Here is the wiki page with difference overview:
* https://github.com/ruslo/hunter/wiki/example.find_package.config

**Comment by alamaison** [comment]
On Fri, 15 Jan 2016 20:01 Richard Thomson <notifications@github.com> wrote:

> I'm not sure how CONFIG affects the whole setup, but you're right I
> didn't have that.
>
The other day you were adding about the double-colon style namespaceing. In
a way, CONFIG activates that kind of package lookup.

CMake has two styles of package lookup. The old way, otherwise known as
module lookup, relies on a FindXXX.CMake module shipped with CMake or
included with the client project.

The newer way, CONFIG lookup, expects the package to package to provide
it's own description of where to find itself in an XXXConfig.CMake file
installed to a well known location on the search path. Combined with
IMPORTED targets and INTERFACE properties, it gives CMake a really powerful
mechanism for handling dependencies. It really simplifies CMake.

Hunter works with both method, but can only really be zero-config for
client projects if using the config method.

**Comment by caseymcc** [comment]
I am modifying hunter for boost to allow for shared compiles of the libraries. I have it working but have a couple of questions though;

1. Should we be keying off BUILD_SHARED_LIBS or Boost_USE_STATIC_LIBS
2. If we key off BUILD_SHARED_LIBS, should we default MSVC to BOOST_ALL_DYN_LINK=ON
3. If the cmake args for Boost are expected to be in the config.cmake/defualt.cmake files it feels like we should not need any boost definitions in the CMakeLists.txt, however BOOST_ALL_DYN_LINK/BOOST_ALL_NO_LIB(for MSVC)/and maybe a few more needs to be set for the boost headers to properly include the libs. Should this be done in the FindBoost.cmake with add_definition or are we expecting the user to redefine them in the CMakeLists.txt?

**Comment by ruslo** [comment]
For the record: https://github.com/ruslo/hunter/pull/550
